### ISSUE TITLE
H3: body buffer size configuration

### ISSUE THREAD

**Issue by stammw** [djc/quinn#355]
Make it possible for the user to specify body initial buffer size and max size for receiving it. Replacing the static values at `quinn-h3/src/server.rs:296`, and `quinn-h3/src/client.rs:320` with fields settable by the user.

Receiving the body from the user code would then look like: 

```rust
response.set_max_capacity(n);
response.set_capacity(m);
response.body().and_then(...)
```

**Comment by TimonPost** [comment]
The lines pointed in this issue don't exist anymore. Is this issue still valid?

**Comment by stammw** [comment]
The initial concern was about removing `body::ReadToEnd::new()` params
`capacity` and `size_limit` and offer a request / connection / endpoint level
configuration interface to set those.

`body::ReadToEnd` will soon be removed in an effort to simplify the public API,
I started to draft it in #518. I'm currently studying `hyper::h2` API and some
crates I'd like to see `quinn-h3` integrated into, namely `hyper` and `actix-web`.

Still, there is some work to do with Body chunks size management, even if it is
less trivial than I first thought:

Bodies are transported via an arbitrary number of `DataFrame`s, with a header
indicating the size of the payload for each of them. But what if this `Data`
frame size is `1G`? We don't want to grow a buffer until the whole payload is
received, because it would be too easy for a malicious peer to make our peer go
out of memory.

I came up with a quick solution via `proto::frame::PartialData`, which cuts big
`DataFrame`s into smaller ones, then permits `frame::FrameStream` to yield their
payloads to the user. The problem is that the largest size for a chunk is now
determined by the internal buffer size of `tokio_codec::FramedRead`.

If the buffer is too small, the throughput performance will be limited. We also
want to limit it's growth to prevent it from draining the memory, and having a
way to make it grow by user-determined setting would be a nice feature.

For example, a `content-length` header could be present in the response and we
would like to let the user choose how he wants to handle the body.

So now the issue is more about: what should we do to maximize throughput and let
the user influence the memory management via some settings. Keeping in mind that
we want to stay close to what `h2` does so `quinn-h3` integrates nicely with
crates using it.

Definitely not a trivial issue, but it could bring a good grasp on what's this
project problematics are about.

@djc @Ralith Maybe you have some thoughts about all this?

**Comment by djc** [comment]
How do `DataFrame`s relate to QUIC packets? I would assume a `DataFrame` will always be contained within a QUIC packet. QUIC packets can not be longer than 65536 bytes, so we can simply use that as a cutoff.

Basically, I think we should deliver data frames as they come in, except in order. The interface ought to be similar to what hyper does; I think it currently uses a `HttpBody` trait, which even lives in a separate crate, so we could reuse that? hyper then provides a simple default implementation.

**Comment by stammw** [comment]
I'll try to summarize what the current API does in current state: 

`DataFrame` can be of any size.

Writing any size with `BodyWriter` will result in a `DataFrame` of that size.

And reading on `BodyReader` will yield data as soon as: 
* A complete `DataFrame` has been received 
* A `DataFrame` has been received partially (~ whole RecvStream buffer has been consumed)
* Data from the previous Partial frame has been recieved (any size)
* If called through AsyncRead
  * when the passed buffer is full
  * when a previous "buffer is full" cut-off left some data (can be very small)

So there is two ways of reading:
* `impl AsyncRead for BodyReader`, tries to fill out the passed buffer. Will _copy data twice_, from
  `RecvStream` to `FramedRead`, then from `DataFrame` into the user's buffer.
* `async fn BodyReader::data() -> Result<Bytes>` will return either the first full frame, or all 
  the available data from `FrameStream`. Will copy data once. From `RecvStream`
  to `FrameStream`.
  
`FrameStream` uses an internal buffer to decode from `RecvStream` via `
AsyncRead`. So this copy  seem inevitable, but maybe it is simply also required
for Quic ordered read? Never the less this internal buffer is `8k`  at start and
seem to grow without limit.

Now that the home-made `body::ReadToEnd` as been abandoned, only streaming API
is exposed, so we don't have to deal with a "maximum body size" limit. Still, we
have to take care of this `FramedRead` internal buffer.

Seems like fn `framed_read2_with_buffer<T>(inner: T, mut buf: BytesMut) ->
FramedRead2<T>` offers to pass an initial buffer in. I'm going try to use this
with the desired size.


Regarding the `HttpBody` type in `hyper`, it looks more like an abstraction over
`h1 / h2 / h3(soon)` body handling. `h2` doesn't use it at all.

**Comment by stammw** [comment]
Doesn't look like it's possible to influence this buffer's size. I opened an Issue for this here: https://github.com/tokio-rs/tokio/issues/2156

**Comment by stammw** [comment]
Fixed by #662
